plant item 클릭 시 상세 페이지로 이동 구현

june0122 opened the pull request:
<p align = 'center'>
<img width = '300' src = 'https://user-images.githubusercontent.com/39554623/153594919-84eea48a-a4e8-4389-83e1-4fc7f35e146e.png'>
<img width = '300' src = 'https://user-images.githubusercontent.com/39554623/153594016-03b51363-051d-403d-bf5b-097f8584f14b.png'>
</p>

CollapsingToolbarLayout 이용해서 상세 페이지에서 위로 드래그하면 줄어드는 툴바 적용하려다가 오랜만에 하다보니 조금 꼬여버려서 시간이 잡아먹히고 있습니다.

우선 plant item 클릭 시 상세 페이지로 이동은 구현하였고 coil이나 glide 이용해서 url을 통해 이미지 로딩하는 것만 구현하면 첫 번째 요구사항에서 중요한 부분은 다 구현할 것 같습니다.

june0122 committed:
resource: 아이콘 및 색상값 리소스 추가

june0122 committed:
chore: 메니페스트 NoActionBar 테마 변경

june0122 committed:
refactor: 식물 상세 페이지용 이미지 URL 및 설명 프로퍼티 추가

june0122 committed:
resource: dimens.xml 정의 및 데이터 추가

june0122 committed:
style: 식물 이름 텍스트와 백그라운드의 색상 변경

june0122 committed:
style: 식물 상세 페이지의 테스트용 레이아웃 정의

june0122 committed:
feat: 리사이클러뷰의 아이템 클릭 시 상세 페이지로 이동 구현

june0122 committed:
chore: gradle 및 라이브러리 버전 업데이트

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -9,16 +9,15 @@ import androidx.recyclerview.widget.GridLayoutManager
-
     companion object {
-        fun newInstance() = PlantListFragment()
+        fun newInstance() = PlantListFragment() // 동반 객체의 newInstance() 사용 시의 이점은?
```
Fragment 내에서 전달 받아야 할 값들이 있을 경우 이점이 있지만 newInstance()만 하는 경우는 그 장점이 보이지 않죠.

june0122 replied:
아하 모종의 이유로 프래그먼트가 재생성이 될 경우에 인자가 없는 기본 생성자로 프래그먼트를 재생성하기 때문에 인자 유지를 위해서 `newInstance()`를 사용하는 것이었군요.

관련 글들을 찾아보니 androidx부터는 FragmentFactory를 사용하는 방법에 대해 많이 소개가 되어 있는데, 그렇다면 이제 인자가 있는 프래그먼트를 사용할 경우에 `newInstance()`보다 FragmentFactory를 사용하는게 더 좋은 방법인가요?

#### 참고글
-  [Android] AndroidX의 FragmentFactory 사용하기 : https://zion830.tistory.com/85

theo-f-lab replied:
FragmnetFactory를 사용하는것도 하나의 방법이긴 하지만 요즘은 Navigation 을 함께 사용하다보니 이것 역시 많이 쓰여지지는 않을것 같습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -34,5 +33,26 @@ class PlantListFragment : Fragment() {
         }
         plantListLayoutManager.orientation = GridLayoutManager.VERTICAL
+
+        recyclerView.addOnItemTouchListener(
```
TouchListener를 활용한 단순 아이템 클릭에서 활용하기에는 덩치도있고, 매번 정의해야할게 많을것 같습니다.
이런 경우에는 Adapter 초기화 시 Higher order function 또는 Listener를 넘기고, 그 리스너를 통해 값을 전달 받을 수 있도록 하는게 좋습니다.

```
class RecyclerViewAdapter(val listener: OnItemClickListener) : /* 생략 */ {}
```

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -34,5 +33,26 @@ class PlantListFragment : Fragment() {
+                        val plantData = plantListAdapter.items[position]
+
+                        plantDetailFragment.receivePlantData(plantData)
+                        activity?.supportFragmentManager
```
Fragment는 Activity가 없을 경우 동작하지 않아야해서 이 경우 requireActivity()를 활용하셔도 괜찮습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -34,5 +33,26 @@ class PlantListFragment : Fragment() {
+                        val plantDetailFragment = PlantDetailFragment.newInstance()
+                        val plantData = plantListAdapter.items[position]
+
+                        plantDetailFragment.receivePlantData(plantData)
```
이 부분은 값을 직접적으로 초기화는것이 아닌 Fragment에 argument 던지는 방법으로 변경해주세요.
ex)
PlantDetailFragmnet.newInstance(plantData)
이와 같이 명시해주시고, newInstnace()에서 argument 사용하는 방법으로 변경하세요.

https://stackoverflow.com/questions/9245408/best-practice-for-instantiating-a-new-android-fragment

theo-f-lab commented on app/src/main/res/layout/fragment_plant_detail.xml:
```diff
@@ -0,0 +1,127 @@
+    <com.google.android.material.appbar.AppBarLayout
+        android:id="@+id/app_bar"
+        android:layout_width="match_parent"
+        android:layout_height="240dp"
```
AppBar의 높이를 240으로 한 이유가 있을까요?

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantDetailFragment.kt:
```diff
@@ -4,14 +4,33 @@ import android.os.Bundle
         super.onViewCreated(view, savedInstanceState)
+
+        val plantName = view.findViewById<TextView>(R.id.plantNameTextView)
+        plantName.text = data.name
```
변수를 1번만 사용하는 경우라면 아래와 같이 쓰는 편이 좋습니다.

```
view.findViewById<TextView>(R.id.plantNameTextView).run {
    text = data.name
}

june0122 replied:
run은 lambda result를 반환하지만 apply는 context object를 반환하기 때문에 초기화한 객체를 변수에 할당할 때 많이 사용하는 것으로 기억하고 있습니다.

위의 코드에서 run 대신 apply를 사용해도 동작에는 이상이 없겠지만 내부적으론 불필요한 객체를 반환하게 되는 것이므로 괜히 GC가 제거할 대상만 불필요하게 늘리게 된다고 이해해도 될까요?

theo-f-lab replied:
run/apply를 사용하셔도 사실 GC 제거할 대상이 더 늘어나거나 하진 않아요. 리턴을 람다 표현식의 마지막을 해주는것이지, 있던 GC 대상이 사라지지 않습니다.

객체.run {} 풀어 쓰면 전달 받은 객체도 있고, run 블락 안에 포함되어있는 GC 대상도 그대로입니다.
객체.apply {} 생성과 동시에 객체의 내부 값을 채워넣고, 객체를 리턴하는데 이역시 객체를 그대로 리틴합니다.
결국 접근의 제한을 apply로 해서 객체를 다시 사용할 수 있도록 할것인가.
접근에 제한은 run으로해서 가지고온 객체 대신 다른걸 리턴할것인가의 차이이지 GC 대상이 줄어들거나, 늘어나지 않습니다.

```
val plantName = view.findViewById<TextView>(R.id.plantNameTextView)
plantName.text = data.name
```
-> plantName 객체,
-> plantName 객체의 text 값 변경

```
view.findViewById<TextView>(R.id.plantNameTextView).run {
  text = data.name
}
```
setText한 결과를 run의 함수 블락 밖으로 내어주는것이지 plantName 객체가 만들어져서 사용되는것은 동일합니다. 그러니 불필요한 GC 대상을 늘리는게 아닌 원래 있는거라서 아무상관 없는것이죠.

힌트) inline 키워드

그리고 위 코드는 run도 쓰지 않고 아래와 같이 하는것과 동일하기에 

```
view.findViewById<TextView>(R.id.plantNameTextView).text = data.name
```

이 코드가 run{} 활용한 것과 차이가 없습니다. 그러니 GC는 동일한것이겠죠.

theo-f-lab commented on app/src/main/res/layout/fragment_plant_detail.xml:
```diff
@@ -0,0 +1,127 @@
+                    android:layout_height="match_parent">
+
+                    <ImageView
+                        android:id="@+id/detailBackButton"
```
xml의 view id 작성은 snake case를 활용합니다.
그리고 image, button 등을 구분하기 쉽게 앞에 prefix로 `btn_ `, `img_`, `tv_`, `et_` 등을 붙여 사용합니다.

https://stackoverflow.com/questions/1832290/android-id-naming-convention-lower-case-with-underscore-vs-camel-case

theo-f-lab commented on app/src/main/res/layout/fragment_plant_detail.xml:
```diff
@@ -0,0 +1,127 @@
+                android:layout_marginTop="24dp"
+                android:layout_marginEnd="24dp"
+                android:lineSpacingExtra="2sp"
+                android:text="The sunflower (Helianthus annuus) is a living annual plant in the family Asteraceae, with a large flower head (capitulum). The stem of the flower can grow up to 3 metres tall, with a flower head that can be 30 cm wide. Other types of sunflowers include the California Royal Sunflower, which has a burgundy (red + purple) flower head.  The flower head is actually an inf ...
```
기본값이 아닌 확인을 하는 경우에는 `tools:text=`를 활용할 수 있습니다. 습관 들여두시면 추후 실수를 방지할 수 있습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/utils/ListItemClickListener.kt:
```diff
@@ -0,0 +1,39 @@
+        if (childView != null && listener != null && mGestureDetector.onTouchEvent(e)) {
+            try {
+                listener.onItemClick(childView, view.getChildAdapterPosition(childView))
+            } catch (e: Exception) {
```
이 부분에서 try/catch를 달아둔 이유가 무엇일까요?

june0122 replied:
스택오버플로우와 블로그 글들을 참고하면서 작성한 코드인데, `listener.onItemClick(...)`의 인자로 사용되는 childView가 주어진 좌표 아래의 최상위 뷰이다보니 터치 시의 좌표와 관련된 코드라 에러가 날 가능성이 있어서 try/catch를 달아놓은건가?하고 가볍게 넘어갔었습니다. 

질문을 주셔서 다시 코드를 보니 위에서 childView가 null인지 다 체크하고 들어가는데 굳이 또 필요한가라는 생각이 드네요.

theo-f-lab replied:
try/catch를 달아둘때는 고민할 부분들이 많습니다. 주의해서 달아주시는게 좋고, 차라리 없는 편이 좋습니다. 어떠한 문제가 발생했을지를 예측하지 못한 상태로 달아주면 수정 가능성을 없앤다는 뜻이기 때문이죠.

최소한 오류 레포트로 받아보고, 아 이런 문제가 있으니 그거에 따른  exception을 명확하게 달아주는게 좋습니다. 또는 해결하시거나요. 가장 많이 하는 실수이자 귀찮음에 의해 생성되는 try/catch의 e: Exception은 사실 매우 위험한 존재입니다.

적용하시더라도 추론 또는 이유를 알고 사용하시는게 좋습니다.

-----
적어주신걸 기준으로 여기서는 adapterPostion을 가져오다가 childView가 없을 경우 exception이 발생하거나, adapter 포지션이 잘못되어 List사이즈보다 크거나, 작은 -1 또는 list.size + 1이 넘어올 수 있는 경우를 대응하기 위함일 수 있습니다.
이런 경우는 0에서 list 사이즈 사이의 값일 때만 호출되도록 하는게 방법이겠죠.

또 다른 문제로 list를 비동기로 처리하는 중에 사용자가 찰나의 순간에 아이템을 눌렀거나, 위치 이동을 했거나 등일 수 있습니다.
언제든 발생할 수 있는 오류입니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/adapter/PlantListAdapter.kt:
```diff
@@ -8,23 +8,79 @@ import com.june0122.sunflower.model.data.Plant
-        Plant(image = "", name = "Apple"),
-        Plant(image = "", name = "Apple"),
-        Plant(image = "", name = "Apple"),
+    var items = arrayListOf(
```
이 아이템은 외부에서 주입할 수 있도록 변경해주세요.
변경 시 다음 조건을 지켜주세요.
- item은 언제든 add, remove, clear 할 수 있어야 합니다.
- item 리스트 변수는 private으로 선언해주세요.
- item이 추가되었을 경우 리스트 갱신이 동작하도록 해주세요.
   - 이걸 테스트 하기 위한 addButton을 Fragment에 추가하세요.

june0122 committed:
chore: xml의 view id를 camel case에서 snake case로 수정

image, button 등을 구분하기 쉽게 앞에 prefix로 btn_ , img_, tv_, rv_ 등을 붙여 사용

june0122 committed:
refactor: 인자가 없는 fragment이므로 newInstance() 함수 제거

프래그먼트가 재생성이 될 경우에 인자가 없는 기본 생성자로 프래그먼트를 재생성하기 때문에
Fragment 내에서 전달 받아야 할 값들이 있을 경우에는 newInstance()가 이점이 있다.

june0122 committed:
chore: 확인을 위한 텍스트이므로 `android:text` 대신 `tools:text`를 사용

june0122 committed:
refactor: Plant 객체를 Bundle 객체에 저장하기 위해 Serializable을 구현

june0122 committed:
refactor: 값을 직접적으로 초기화하지 않고 Fragment에 argument를 던지는 방법으로 변경

receivePlantData() 메서드를 통해 직접적으로 값을 초기화하지 않고
PlantDetailFragment.newInstance(plantData)와 같이
Fragment에 argument를 던지는 방법으로 변경

june0122 committed:
refactor: Fragment는 Activity가 없을 경우 동작하지 않아야하므로 `requireActivity()` 사용

june0122 committed:
refactor: TouchListener 대신 간략히 정의한 listener 인터페이스로 리사이클러뷰 클릭 인벤트 구현

기존의 TouchListener를 단순한 아이템 클릭에 활용하기엔 덩치도 있고, 정의할 내용이 많기 때문에
Adapter 초기화 시에 생성자를 통해 Listener를 넘기고, 그 리스너를 통해 값을 전달 받는게 좋다.

june0122 committed:
refactor: PlantListFragment는 전달 받을 값이 없으므로 `newInstance()` 제거

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/model/data/Data.kt:
```diff
@@ -1,3 +1,9 @@
+    val imageUrl: String,
+    val name: String,
+    val description: String,
+    ): Serializable
```
Serializable 하셔도 좋은 방법이지만 Android에서 기본 제공하는 Parcelable을 활용하는게 좋습니다.

아래 자료를 참고하세요.

Parcelable 공식 문서
https://developer.android.com/reference/android/os/Parcelable

Kotlin에서 Parcelable 사용하는 방법
https://developer.android.com/kotlin/parcelize

Serializable과 Parcelable 비교
https://stackoverflow.com/questions/39518268/when-and-how-to-implement-parcelable-vs-serializable#:~:text=Per%20the%20google%20engineers%20Parcelable,objects%20in%20comparison%20to%20Serialization.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/adapter/PlantListAdapter.kt:
```diff
@@ -35,8 +92,9 @@ class PlantListAdapter : RecyclerView.Adapter<PlantListViewHolder>() {
-            val plantName = findViewById<TextView>(R.id.plantNameTextView)
+            val plantName = findViewById<TextView>(R.id.tv_plant_name)
             plantName.text = model.name
+            setOnClickListener { listener.onPlantSelected(position) }
```
setOnClickListener의 경우 onCreateViewHolder와 onBindViewHolder 중 어느 위치에서 설정하는것이 좋을까요? 직접 로그 출력을 통해 확인해보세요.

힌트 1) 매번 호출되는것과, 필요한 시점에 호출되는것의 차이가 있습니다.
힌트 2) getAbsoluteAdapterPosition()라는 게 있습니다.(과거엔 getAdapterPosition())
https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.ViewHolder#getAbsoluteAdapterPosition()

june0122 replied:
#### A1
`onCreateViewHolder`는 viewHolder를 화면에 보이는 항목의 개수(+ 위아래 버퍼)만큼만 생성한 뒤에 더 이상 호출되지 않는 반면, `onBindViewHolder`는 뷰홀더 객체가 재사용될 경우 데이터를 업데이트하므로 스크롤을 해서 데이터가 새롭게 바인딩될 때마다 호출되니 `onCreateViewHolder` 함수 내부에 setOnClickListener를 설정하는 것이 좋을 것 같습니다.

#### A2
`getAbsoluteAdapterPosition()`은 문서를 읽어봤지만 왜 힌트로 주셨는지 연관성을 못찾겠습니다... 🥲

#### A3
`getAbsoluteAdapterPosition()` 메서드는 앱의 build.gradle에 리사이클러뷰에 대한 의존성을 따로 추가해야 사용이 가능하더군요. 평상시엔 따로 리사이클러뷰를 추가하지 않아도 관련된 기능들을 구현하는데 문제가 없었던 것 같은데 의존성에 추가해주는게 원래 옳은 방법인가요?

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantDetailFragment.kt:
```diff
@@ -4,14 +4,46 @@ import android.os.Bundle
+    companion object {
+        fun newInstance(plantData: Plant): PlantDetailFragment {
+            val args = Bundle().apply {
+                putSerializable(PLANT_DATA, plantData)
```
👍

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantDetailFragment.kt:
```diff
@@ -4,14 +4,46 @@ import android.os.Bundle
+
+    companion object {
+        fun newInstance(plantData: Plant): PlantDetailFragment {
+            val args = Bundle().apply {
```
Bundle().apply {}를 사용하는것도 좋지만 
bundleOf를 활용할 수 있습니다.

https://stackoverflow.com/questions/49342909/kotlin-extensions-for-android-how-to-use-bundleof

core-ktx 라이브러리를 추가하면 사용 가능하니 문서를 참고하세요.
https://developer.android.com/kotlin/ktx

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/utils/PlantSelectedListener.kt:
```diff
@@ -0,0 +1,5 @@
+package com.june0122.sunflower.utils
+
+interface PlantSelectedListener {
```
SAM을 기억하신다면 어떠한 차이가 있는지 직접 적용해보세요.

https://kotlinlang.org/docs/fun-interfaces.html

june0122 replied:
코멘트보자마자 이마를 탁 쳤네요 ㅎㅎ SAM 인터페이스인데 `fun interface`로 정의할 생각을 안하다니...!

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -9,30 +9,40 @@ import androidx.recyclerview.widget.GridLayoutManager
-        fun newInstance() = PlantListFragment()
-    }
+    // onCreateView(...)와 onViewCreated(...) 중 어디에 리사이클러뷰 및 어댑터 초기화를 해주는게 적절한가
+    // ViewHolder나 Adapter를 다른 클래스 파일로 분리하지 않고 Fragment의 inner class로 선언하는 경우도 있던데 어떤 방법이 적절할까?
```
inner class로 선언하는것과 nested class로 선언하는것 어떤 차이가 있을까요? 과거 자바/코틀린을 학습하면서 여러번 이야기 나누었던 내용입니다.

```
inner class XXX

class XXX
```

힌트) fragment 안에서만 활용된다고 inner class를 작성한다면 RecyclerView가 Fragment에 모두 접근할 수 있겠죠.
힌트2) nested class를 작성한다면 어떤 차이가 있을까요?
정답이 있는건 아니지만 소소하게 알고 넘어가야할 부분들이 있겠죠. 파일로 나눴을때의 장점은 한 파일에서 딱 필요한 것들만 모일 수 잇는 반면, 
동일하게 필요하다고해서 Fragment 내부에 class들을 모아두면 해석하기 어려워질 수 있겠죠. 현재는 코드가 짧아서 문제가 없겠으나 100줄 200줄... 1,000줄이 넘어가면

june0122 replied:
자바는 static nested class(정적 중첩 클래스)를 구현하려면 `static` 키워드를 붙여야 하지만, 코틀린은 자바와 다르게 아무 키워드를 붙이지 않을 경우 static nested class 구현이 됩니다. non-static한 nested class 구현을 위해선 `inner` 키워드를 붙여야 하고요.

nested class로 ViewHolder나 Adapter를 작성한다면 Fragment의 필드에 마음대로 접근할 수 있는 문제는 사라지지만 말씀하신대로 Fragment의 내용이 많아지게 되면 가독성이 떨어질 위험이 크겠군요..!

june0122 committed:
refactor: PlantSelectedListener가 SAM 인터페이스(함수형 인터페이스)이므로 `fun interface`로 정의

june0122 committed:
chore: Parcelable 지원 포함을 위해 앱의 build.gradle 파일에 `kotlin-parcelize` 플러그인 추가

june0122 committed:
refactor: 다른 프래그먼트에 객체 전달을 위해 Serializable 대신 Parcelable을 사용

june0122 committed:
chore: 리사이클러뷰 라이브러리 추가

june0122 committed:
refactor: `setOnClickListener` 호출 위치 변경

onCreateViewHolder는 viewHolder를 화면에 보이는 항목의 개수(+ 위아래 버퍼)만큼만 생성한 뒤에 더 이상 호출되지 않는 반면,
onBindViewHolder는 뷰홀더 객체가 재사용될 경우 데이터를 업데이트하므로 스크롤을 해서 데이터가 새롭게 바인딩될 때마다 호출되므로
onCreateViewHolder 함수 내부에 setOnClickListener를 설정

deprecated된 getAdapter() 대신 getAbsoluteAdapterPosition()으로 변경

june0122 committed:
refactor: `Bundle().apply { ... }` 대신 `bundleOf()` 활용

june0122 committed:
chore: 매니페스트에 인터넷 퍼미션 추가

june0122 committed:
chore: 앱의 build.gradle에 Coil 이미지 로딩 라이브러리 의존성 추가

june0122 committed:
style: CollapsingToolbarLayout 내부에 이미지뷰 추가

june0122 committed:
style: 상하 margin 추가

june0122 committed:
feat: 리사이클러뷰와 상세 페이지에 url을 통해 이미지 추가
